### PR TITLE
Make overlap more flexible

### PR DESCRIPTION
**Pull request by DiegoDeGusem**
With this PR it should now be possible to have overlapping grids without overlapping geometries.

The function is added to `MumaxWorld` because Grid has no access to geometries. It still uses the old `Grid::overlaps` function as a quick first check to see if the grids overlap. If they do overlap it uses a GPU function in `reduce.cu` to check if geometry cells overlap.

Two tests were also added to `test_geometry`, one where the geometries don't overlap and one where they do.

### CODE REVIEW

**Review by JonathanMaes — APPROVED** [review]

File: src/physics/mumaxworld.cpp [hunk 1]
```diff
@@ -369,4 +370,36 @@ void MumaxWorld::centerDomainWall(int comp, int axis) {
...
+
+  int3 o1 = grid1.origin(), s1 = grid1.size();
+  int3 o2 = grid2.origin(), s2 = grid2.size();
+
+  // Overlapping box
+  int3 lo{std::max(o1.x, o2.x), std::max(o1.y, o2.y), std::max(o1.z, o2.z)};
+  int3 hi{std::min(o1.x + s1.x, o2.x + s2.x),
+          std::min(o1.y + s1.y, o2.y + s2.y),
+          std::min(o1.z + s1.z, o2.z + s2.z)};
+  int3 n{hi.x - lo.x, hi.y - lo.y, hi.z - lo.z};
```
Comment (lamoreel) [inline src/physics/mumaxworld.cpp:399]: Perhaps more meaningful variable names are appropriate here.

Comment (lamoreel) [reply src/physics/mumaxworld.cpp:399]: (same comment in `reduce.cu` where even less context is given.)

File: src/core/reduce.cu [hunk 2]
```diff
@@ -289,4 +289,56 @@ bool isUniformField(const Field& f) {
...
+  bool found = false;
+  for (int i = tid; i < ncells; i += BLOCKDIM) {
+    // unravel flat index i into a 3D offset within [0, n)
+    int x = i % n.x;
+    int y = (i / n.x) % n.y;
+    int z = i / (n.x * n.y);
+    int3 coo{lo.x + x, lo.y + y, lo.z + z};
+
+    bool in1 = !geometry1 || geometry1[grid1.coord2index(coo)];
+    bool in2 = !geometry2 || geometry2[grid2.coord2index(coo)];
```
Comment (lamoreel) [inline src/core/reduce.cu:310]: What's the reason behind the first condition `!geometry1`(2)? Why isn't the condition after the `||` sufficient?

Comment (lamoreel) [inline src/core/reduce.cu:310]: If I understand correctly, `coo` is calculated from a cell within a minimal bounding box encapsulating both grids. What if `coo` is not inside a `grid`? Is this still safe?

File: test/test_geometry.py [hunk 3]
```diff
@@ -52,3 +53,44 @@ def test_invalid_function_geometry(self):
...
+        plane = Cuboid(planeside, planeside, cs).translate(6*cs, 6*cs, 5*cs)
+        diff = bigCube - smallCube
+
+        magnet1 = Ferromagnet(world, grid, geometry=diff)
+        magnet2 = Ferromagnet(world, grid, geometry=plane) # or Ferromagnet(world, Grid((4, 4, 1), origin=(6, 6, 5))
+
+        assert magnet1 is not None
+        assert magnet2 is not None
+
+    def test_overlap_fail(self):
```
Comment (lamoreel) [inline test/test_geometry.py:77]: I know I gave this example a while ago as a case that should (not) be accepted as a valid geometry, but I guess in this case something like this would suffice:
```python
magnet1 = Ferromagnet(world, grid)
magnet2 = Ferromagnet(world, grid, geometry=np.zeros(grid.shape)) # or np.ones()
```
Perhaps it's also meaningful to write a test where the magnets only overlap at one corner (by shifting the origin), as to isolate the new functionality.

**Review by lamoreel — CHANGES_REQUESTED** [review]
Glad to see my suggestion from a while back into a PR 😊 !

File: src/core/reduce.cu (hunk 2 above)
Comment (JonathanMaes) [reply src/core/reduce.cu:310]: I believe OR-statements are executed from left to right.
So if `geometry1` does not exist (or in any way results in `!geometry1` = `true`), the second half of the OR-statement is not evaluated, which in that case would lead to an indexing error.

Comment (JonathanMaes) [reply src/core/reduce.cu:310]: I think it's the intersection of both grids (see `mumaxworld.cpp` line 394+), so `coo` should always be inside both grids.

Comment (lamoreel) [reply src/core/reduce.cu:310]: My point exactly 😊 . The second half should be sufficient, since a global geometry check is performed in `mumaxworld.cpp` line 386 onward.

File: src/physics/mumaxworld.hpp [hunk 4]
```diff
@@ -38,7 +38,7 @@ class MumaxWorld : public World {
   /** Uniform bias magnetic field which will affect all magnets in the world. */
   real3 biasMagneticField;
 
-  void checkAddibility(Grid grid, std::string name) const;
+  void checkAddibility(Grid grid, GpuBuffer<bool> geometry, std::string name) const;
```
Comment (ilateur) [inline src/physics/mumaxworld.hpp]: I think it is best here to pass geometry by constant reference `const GpuBuffer<bool>& geometry1`. Then it's not copied and unmodified.

File: src/physics/mumaxworld.cpp [hunk 5]
```diff
@@ -369,4 +370,40 @@ void MumaxWorld::centerDomainWall(int comp, int axis) {
...
+  int3 overlapMaximum{std::min(o1.x + s1.x, o2.x + s2.x),
+          std::min(o1.y + s1.y, o2.y + s2.y),
+          std::min(o1.z + s1.z, o2.z + s2.z)};
+  int3 overlapCells{overlapMaximum.x - overlapOrigin.x,
+         overlapMaximum.y - overlapOrigin.y,
+         overlapMaximum.z - overlapOrigin.z};
+
+  // Check all cells in the overlapping box on GPU
+  return geometriesOverlap(grid1, hasGeo1 ? geometry1.get() : nullptr,
+                           grid2, hasGeo2 ? geometry2.get() : nullptr,
```
Comment (ilateur) [inline src/physics/mumaxworld.cpp]: I think `GpuBuffer` already returns `nullptr` if `size` is 0, so these ternary operators are unnecessary.

File: src/physics/mumaxworld.cpp [hunk 6]
```diff
@@ -369,4 +370,40 @@ void MumaxWorld::centerDomainWall(int comp, int axis) {
...
+                           overlapOrigin, overlapCells);
```
Comment (ilateur) [inline src/physics/mumaxworld.cpp]: I think it would be cleaner to combine `overlapOrigin` and `overlapCells` into an `overlapGrid` and pass that as an argument.

File: src/core/reduce.cu [hunk 7]
```diff
@@ -289,4 +289,59 @@ bool isUniformField(const Field& f) {
...
+}
+
+bool geometriesOverlap(Grid grid1, bool const* geometry1,
+                       Grid grid2, bool const* geometry2,
+                       int3 overlapOrigin, int3 overlapCells) {
+  int ncells = overlapCells.x * overlapCells.y * overlapCells.z;
+
+  GpuBuffer<bool> d_result(1);
+  cudaLaunchReductionKernel(k_geometriesOverlap, d_result.get(),
+                            grid1, geometry1, grid2, geometry2, overlapOrigin, overlapCells, ncells);
```
Comment (ilateur) [inline src/core/reduce.cu]: If `overlapOrigin` and `overlapCells` are combined into an `overlapGrid` then `ncells` can be asked from that grid, either here or inside the reduction kernel.

File: src/core/reduce.cu [hunk 8]
```diff
@@ -289,4 +289,59 @@ bool isUniformField(const Field& f) {
...
+  __shared__ bool sdata[BLOCKDIM];
+  int tid = threadIdx.x;
+
+  bool found = false;
+  for (int i = tid; i < ncells; i += BLOCKDIM) {
+    // unravel flat index i into a 3D offset within [0, n)
+    int x = i % overlapCells.x;
+    int y = (i / overlapCells.x) % overlapCells.y;
+    int z = i / (overlapCells.x * overlapCells.y);
+    int3 coo{overlapOrigin.x + x, overlapOrigin.y + y, overlapOrigin.z + z};
```
Comment (ilateur) [inline src/core/reduce.cu]: `Grid` already has `index2coord` doing exactly this.

File: src/core/reduce.hpp [hunk 9]
```diff
@@ -13,4 +14,7 @@ real fieldComponentAverage(const Field&, int comp);
 std::vector<real> fieldAverage(const Field&);
 bool idxInRegions(GpuBuffer<unsigned int>, unsigned int idx);
 bool isUniformFieldComponent(const Field&, int);
-bool isUniformField(const Field&);
\ No newline at end of file
+bool isUniformField(const Field&);
+bool geometriesOverlap(Grid grid1, bool const* geometry1,
+                       Grid grid2, bool const* geometry2,
+                       int3 lo, int3 n);
```
Comment (ilateur) [inline src/core/reduce.hpp]: `lo` and `n` are not very descriptive (but is also solved by passing overlap grid)

File: src/core/reduce.cu (hunk 7 above)
Comment (ilateur) [reply src/core/reduce.cu]: Convention in other reduction kernels seems to be to get `ncells` inside the kernel rather than passing it.

File: test/test_geometry.py [hunk 10]
```diff
@@ -1,5 +1,6 @@
 import numpy as np
 import pytest
+from mumaxplus.util.shape import Cuboid
```
Comment (ilateur) [inline test/test_geometry.py]: `Cuboid` is never used? (sadly)

File: test/test_geometry.py
Comment (ilateur) [inline test/test_geometry.py:1]: It would be interesting to test is less trivial case too, for example two ellipses/ellipsoids that don't touch, but their grids do overlap.

**Review by ilateur — CHANGES_REQUESTED** [review]

File: src/physics/mumaxworld.cpp [hunk 11]
```diff
@@ -24,6 +24,8 @@
 #include "torque.hpp"
 #include "window.hpp"
 
+#include <iostream>
```
Comment (ilateur) [inline src/physics/mumaxworld.cpp]: why?

**Review by ilateur — CHANGES_REQUESTED** [review]

File: src/physics/mumaxworld.cpp (hunk 11 above)
Comment (DiegoDeGusem) [reply src/physics/mumaxworld.cpp]: oops

**Review by ilateur — APPROVED** [review]
